lib/spack/spack/spec.py

```diff
@@ -2896,6 +2906,30 @@ def _mark_concrete(self, value=True):
                 s.clear_cached_hashes()
             s._mark_root_concrete(value)
 
+    def _assign_hashes(self):
+        """Assign package hash and then DAG hash.
+
+        This should be called immediately after concretization.
+
+        We can't compute `package_hash()` lazily, because it needs to be assigned with
+        the package hash *from concretization time*. So, `to_node_dict()` can't just
+        assume that it can compute the `package_hash()` itself -- it needs to either see
+        or not see a `_package_hash` attribute. Old-style specs from *before* DAG hash
+        included the package hash will not have a `_package_hash` at all, and we can't
+        safely recompute it later. New-style concrete specs will have a `_package_hash`
+        assigned at concretization time.
+
+        """
+        for spec in self.traverse():
+            # Already concrete specs either already have a package hash (new dag_hash())
+            # or they never will b/c we can't know it (old dag_hash()). Skip them.
+            if not spec.concrete:
```

Comment:
Isn't this unreachable, since you're calling `Spec._assign_hashes()` immediately after `Spec._mark_concrete`?

Reply:
yes sigh.  Fixing...